Add `THREADS_ENABLED` macro in order to compile Godot to run on the main thread

adamscott opened the pull request:
# Context
This PR adds the `THREADS_ENABLED` macro and the ~~`use_threads=yes|no`~~ `threads=yes|no` build option.

There could be a way to make the `THREADS_ENABLED` modifications run time instead of build time, but the Web platform would still need two separate build targets, one with threads, the other without, due to emscripten limitations.

Has the good side-effect to make web builds compatible with macOS / iOS.

# Demo
https://adamscott.github.io/truck-town-demo-main-thread/
https://adamscott.github.io/truck-town-demo-main-thread-nosound/ (runs without sound, in order to test performance)

![image](https://github.com/godotengine/godot/assets/270928/729b0a93-81dd-4462-85ab-79c492635fb6)

# Progress
## Threads
- [x] Make a [WorkerThreadPool](https://github.com/godotengine/godot/blob/0ca8542329888e8dccba89d59d3b728090c29991/core/object/worker_thread_pool.h#L43) that executes single threaded.

## File system
- [x] Fix issue where file system don't seem to be updated

## Audio
- [x] Port the [Audio Driver](https://github.com/godotengine/godot/blob/3.5/platform/javascript/audio_driver_javascript.h) from Godot 3
- [x] Fix broken driver?!
  - Now only works on HTTPS (AudioWorklet is only available for HTTPS contexts), I should maybe investigate if it's possible to make it work on HTTP.
    - Should be another issue/PR

## Rendering
- [x] RenderingServer should be forced to run single threaded.

## Physics
- [x] PhysicsServer2D/3D should be forced to run single threaded.
- [x] Fix broken physics

## Core
- [x] Add a variable that say if threads are supported or not
  - Added `OS.are_thread_enabled()` function

## Export
- [x] Add `web_compatible` export
  
## Github
- [x] Add CI for `use_threads=no`

## Misc
- [x] Make the [IP Resolver](https://github.com/godotengine/godot/blob/0ca8542329888e8dccba89d59d3b728090c29991/core/io/ip.cpp#L130) work single threaded.
- [x] Make [RaycastOcclusionCull](https://github.com/godotengine/godot/blob/0ca8542329888e8dccba89d59d3b728090c29991/modules/raycast/raycast_occlusion_cull.cpp#L415) work single threaded.

# Fixes
Fixes #85938

# Known issues
- When running in other platforms than the web, sound is broken.

# Edit
## 2024-01-12 
- Updated `use_threads=yes|no` to `threads=yes|no`.
- Added non-web sound known issue

_Production edit: closes godotengine/godot-roadmap#11_

adamscott committed:
Add THREADS_ENABLED macro in order to compile Godot to run on the main thread

akien-mga commented on platform/web/emscripten_helpers.py:
```diff
@@ -115,6 +118,10 @@ def create_template_zip(env, js, wasm, worker, side):
     )
 
 
+def get_template_zip_path(env):
+    return "#bin/.web_zip"
```
I'm assuming the reason to make this a function was so you can use a different path for the two types of builds? This seems to be missing.

adamscott replied:
Initially, it was, that function had logic. But now it doesn't have. But it's better now even if it's returning a static string because there was 2 magic strings in the code otherwise.

akien-mga commented on platform/web/export/export_plugin.cpp:
```diff
@@ -330,6 +330,7 @@ void EditorExportPlatformWeb::get_export_options(List<ExportOption> *r_options)
 	r_options->push_back(ExportOption(PropertyInfo(Variant::STRING, "custom_template/release", PROPERTY_HINT_GLOBAL_FILE, "*.zip"), ""));
 
 	r_options->push_back(ExportOption(PropertyInfo(Variant::BOOL, "variant/extensions_support"), false)); // Export type.
+	r_options->push_back(ExportOption(PropertyInfo(Variant::BOOL, "variant/compatibility_mode"), false)); // Compatibility (ie. run without COEP/COOP headers).
```
```suggestion
	r_options->push_back(ExportOption(PropertyInfo(Variant::BOOL, "variant/compatibility_mode"), false)); // Compatibility (i.e. run without COEP/COOP headers).
```
But same question as I wrote on the docs. Should we put the focus on headers, or on the actual feature provided by Godot (which browsers impose specific header requirements on)?

akien-mga commented on platform/web/export/export_plugin.cpp:
```diff
@@ -372,25 +373,26 @@ bool EditorExportPlatformWeb::has_valid_export_configuration(const Ref<EditorExp
 	// Look for export templates (first official, and if defined custom templates).
-	bool dvalid = exists_export_template(_get_template_name(extensions, true), &err);
-	bool rvalid = exists_export_template(_get_template_name(extensions, false), &err);
+	bool debug_standard_valid = exists_export_template(_get_template_name(extensions, compatibility_mode, true), &err);
```
That rename sounds like you were planning to duplicate the checks before adding it as a bool to the function. So the names can likely stay `dvalid` and `rvalid` like they are for all other platforms.

akien-mga commented on platform/web/export/export_plugin.h:
```diff
@@ -56,11 +56,14 @@ class EditorExportPlatformWeb : public EditorExportPlatform {
 			name += "_dlink";
 		}
+		if (p_compatible) {
+			name += "_compatible";
```
Yeah I'm not fond of that name. Why would one NOT want a "compatible" version?

I think "nothreads" could work.

akien-mga commented on servers/rendering/rendering_server_default.h:
```diff
@@ -78,7 +78,7 @@ class RenderingServerDefault : public RenderingServer {
 	void _thread_loop();
 
-	Thread::ID server_thread;
+	Thread::ID server_thread = Thread::get_caller_id();
```
Is this safe to assign like this? @RandomShaper

RandomShaper replied:
Some servers (including this) have this:
```
	if (!p_create_thread) {
		server_thread = Thread::get_caller_id();
	} else {
		server_thread = 0;
	}
```

I'd say it'd be better to keep the `server_thread = 0;` in the header and remove the `else` in these cases.

RandomShaper replied:
That's a good point. Those `if`s should be replaced by just `server_thread = Thread::get_caller_id()` (or, alternatively, `Thread::get_main_id()`). Regardless, `p_create_thread`, at that point the "server thread" is the main one. The `0` would be only interesting as a default for initialization to avoid the garbage until something more sensible can be assigned. Later, the `thread_loop` functions will reassign `server_thread` and, that time, if the server is actually running on a non-main thread, the right value will be collected there.

adamscott replied:
I just realized that `server_thread` had garbage data in it because I skipped the initialization. It was inside a `THREADS_ENABLED`.

akien-mga commented on core/os/condition_variable.h:
```diff
@@ -31,6 +31,9 @@
 #ifndef CONDITION_VARIABLE_H
 #define CONDITION_VARIABLE_H
 
+#include "core/os/mutex.h"
+#include "core/typedefs.h"
```
Adding these includes seems correct, I wonder how it compiled before? I guess it just happened to be always included after `mutex.h`?

akien-mga commented on core/os/condition_variable.h:
```diff
@@ -63,5 +67,14 @@ class ConditionVariable {
 		condition.notify_all();
 	}
 };
+#else
```
Like I suggested elsewhere, in this situation I'd add empty lines around the `#ifdef THREADS_ENABLED` / `#else` / `#endif // THREADS_ENABLED` to make it easier to see where the two branches start and end.

Same for other similar conditional class definitions.

akien-mga commented on core/os/mutex.h:
```diff
@@ -66,6 +67,11 @@ class MutexImpl {
 	_ALWAYS_INLINE_ bool try_lock() const {
 		return mutex.try_lock();
 	}
+#else
+	_ALWAYS_INLINE_ void lock() const {}
+	_ALWAYS_INLINE_ void unlock() const {}
+	_ALWAYS_INLINE_ bool try_lock() const { return true; }
+#endif
```
I know you went for the most surgical changes to get things to compile first, which is a great approach.

To polish further now that it works, I think we should make intent clear in the API by more explicitly bypassing threading APIs. What I mean is typically (at least for short classes/functions) is to fully enclose the class/function declaration in the macros, and provide a fallback/dummy version for the no-threads branch.

In this case, I would take it as far as just having two branches for everything declared in that file.

```cpp
#ifdef THREADS_ENABLED

#ifdef MINGW_ENABLED
#define MINGW_STDTHREAD_REDUNDANCY_WARNING
#include "thirdparty/mingw-std-threads/mingw.mutex.h"
#define THREADING_NAMESPACE mingw_stdthread
#else
#include <mutex>
#define THREADING_NAMESPACE std
#endif

template <class MutexT>
class MutexLock;

template <class StdMutexT>
class MutexImpl {
	friend class MutexLock<MutexImpl<StdMutexT>>;
	...
};

...

using Mutex = MutexImpl<THREADING_NAMESPACE::recursive_mutex>; // Recursive, for general use
using BinaryMutex = MutexImpl<THREADING_NAMESPACE::mutex>; // Non-recursive, handle with care
extern template class MutexImpl<THREADING_NAMESPACE::recursive_mutex>;
extern template class MutexImpl<THREADING_NAMESPACE::mutex>;
extern template class MutexLock<MutexImpl<THREADING_NAMESPACE::recursive_mutex>>;
extern template class MutexLock<MutexImpl<THREADING_NAMESPACE::mutex>>;

#else // No threads.

template <class MutexT>

class MutexLock;
template <class StdMutexT>

class MutexImpl {
	// stubs
	...
};

...

#endif // THREADS_ENABLED
```

There might be some repetition in the overall structure, but I think it's clearer for the reader to see all these as completely separate implementations of the same interface, instead of Frankeinstein classes where only parts are compile flag specific.

Discussion welcome, CC @Faless

adamscott replied:
I replaced the Frankeinstein cutout with two different parts.

akien-mga reviewed (commented):
Amazing work!

I haven't tested yet, only did a quick code quality review.

adamscott commented on platform/web/doc_classes/EditorExportPlatformWeb.xml:
```diff
@@ -45,6 +45,9 @@
 		</member>
 		<member name="progressive_web_app/orientation" type="int" setter="" getter="">
 		</member>
+		<member name="variant/compatibility_mode" type="bool" setter="" getter="">
+			If enabled, the exported game will be able to run without COEP/COOP headers, but with the drawbacks of not being able to use threads and having some audio issues.
+		</member>
```
Either both of these options or "performance_mode" (the reverse of "compatibility_mode")? Like, why do we use threads after all?

jordo replied:
+1 For being explicit about thread support in the name as it's clear.

AThousandShips commented on core/object/worker_thread_pool.cpp:
```diff
@@ -496,6 +503,10 @@ bool WorkerThreadPool::is_group_task_completed(GroupID p_group) const {
 void WorkerThreadPool::wait_for_group_task_completion(GroupID p_group) {
+#ifndef THREADS_ENABLED
+	return;
+#endif
```
I think this should be an if/else/endif block instead, this would complain on some compilers due to unreachable code

adamscott commented on core/os/semaphore.h:
Will fix this Frankeinstein class in a sec.

jordo commented:
Nice work!  Since this adds a THREADS_ENABLED macro, should probably add a `use_threads=no` CI for this web target?  Feels like this target will need CI to catch commits which don't guard future thread dependencies and have a single threaded impl.

adamscott commented:
> Nice work!  Since this adds a THREADS_ENABLED macro, should probably add a `use_threads=no` CI for this web target?  Feels like this target will need CI to catch commits which don't guard future thread dependencies and have a single threaded impl.

In fact, it should have CI for both. I'll try to add them to the github-actions.

RandomShaper commented on core/os/condition_variable.h:
Let me suggest this alternative to this whole file:
```
/**************************************************************************/
/*  condition_variable.h                                                  */
/**************************************************************************/
/*                         This file is part of:                          */
/*                             GODOT ENGINE                               */
/*                        https://godotengine.org                         */
/**************************************************************************/
/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
/*                                                                        */
/* Permission is hereby granted, free of charge, to any person obtaining  */
/* a copy of this software and associated documentation files (the        */
/* "Software"), to deal in the Software without restriction, including    */
/* without limitation the rights to use, copy, modify, merge, publish,    */
/* distribute, sublicense, and/or sell copies of the Software, and to     */
/* permit persons to whom the Software is furnished to do so, subject to  */
/* the following conditions:                                              */
/*                                                                        */
/* The above copyright notice and this permission notice shall be         */
/* included in all copies or substantial portions of the Software.        */
/*                                                                        */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
/**************************************************************************/

#ifndef CONDITION_VARIABLE_H
#define CONDITION_VARIABLE_H

#ifdef THREADS_ENABLED

#ifdef MINGW_ENABLED
#define MINGW_STDTHREAD_REDUNDANCY_WARNING
#include "thirdparty/mingw-std-threads/mingw.condition_variable.h"
#define THREADING_NAMESPACE mingw_stdthread
#else
#include <condition_variable>
#define THREADING_NAMESPACE std
#endif

// An object one or multiple threads can wait on a be notified by some other.
// Normally, you want to use a semaphore for such scenarios, but when the
// condition is something different than a count being greater than zero
// (which is the built-in logic in a semaphore) or you want to provide your
// own mutex to tie the wait-notify to some other behavior, you need to use this.

class ConditionVariable {
	mutable THREADING_NAMESPACE::condition_variable condition;

public:
	template <class BinaryMutexT>
	_ALWAYS_INLINE_ void wait(const MutexLock<BinaryMutexT> &p_lock) const {
		condition.wait(const_cast<THREADING_NAMESPACE::unique_lock<THREADING_NAMESPACE::mutex> &>(p_lock.lock));
	}

	_ALWAYS_INLINE_ void notify_one() const {
		condition.notify_one();
	}

	_ALWAYS_INLINE_ void notify_all() const {
		condition.notify_all();
	}
};

#else // No threads.

class ConditionVariable {
public:
	template <class BinaryMutexT>
	void wait(const MutexLock<BinaryMutexT> &p_lock) const {}
	void notify_one() const {}
	void notify_all() const {}
};

#endif // THREADS_ENABLED

#endif // CONDITION_VARIABLE_H
```

RandomShaper commented on core/io/ip.cpp:
```diff
@@ -129,6 +129,8 @@ IPAddress IP::resolve_hostname(const String &p_hostname, IP::Type p_type) {
 PackedStringArray IP::resolve_hostname_addresses(const String &p_hostname, Type p_type) {
 	List<IPAddress> res;
+
+#ifdef THREADS_ENABLED
```
This change would skip the resolve cache in the no-threads case. Is that intended? Moreover, I'd say this file should compile and work well with the void implementations of mutex and friends.

adamscott replied:
It was the first file that I worked on and never looked back. So yes, I'll try it with our modifications to mutex and such.

adamscott replied:
Mmm. I'm currently skeptic. We don't currently run the "Thread.start" code, which loops through the IP adresses to resolve. So I'm afraid the code will get stuck.

adamscott replied:
Nevermind. It works!

RandomShaper commented on core/os/semaphore.h:
Let me suggest this alternative to this whole file:
```
/**************************************************************************/
/*  semaphore.h                                                           */
/**************************************************************************/
/*                         This file is part of:                          */
/*                             GODOT ENGINE                               */
/*                        https://godotengine.org                         */
/**************************************************************************/
/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
/*                                                                        */
/* Permission is hereby granted, free of charge, to any person obtaining  */
/* a copy of this software and associated documentation files (the        */
/* "Software"), to deal in the Software without restriction, including    */
/* without limitation the rights to use, copy, modify, merge, publish,    */
/* distribute, sublicense, and/or sell copies of the Software, and to     */
/* permit persons to whom the Software is furnished to do so, subject to  */
/* the following conditions:                                              */
/*                                                                        */
/* The above copyright notice and this permission notice shall be         */
/* included in all copies or substantial portions of the Software.        */
/*                                                                        */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
/**************************************************************************/

#ifndef SEMAPHORE_H
#define SEMAPHORE_H

#ifdef THREADS_ENABLED

#include "core/error/error_list.h"
#include "core/typedefs.h"
#ifdef DEBUG_ENABLED
#include "core/error/error_macros.h"
#endif

#ifdef MINGW_ENABLED
#define MINGW_STDTHREAD_REDUNDANCY_WARNING
#include "thirdparty/mingw-std-threads/mingw.condition_variable.h"
#include "thirdparty/mingw-std-threads/mingw.mutex.h"
#define THREADING_NAMESPACE mingw_stdthread
#else
#include <condition_variable>
#include <mutex>
#define THREADING_NAMESPACE std
#endif

class Semaphore {
private:
	mutable THREADING_NAMESPACE::mutex mutex;
	mutable THREADING_NAMESPACE::condition_variable condition;
	mutable uint32_t count = 0; // Initialized as locked.
#ifdef DEBUG_ENABLED
	mutable uint32_t awaiters = 0;
#endif

public:
	_ALWAYS_INLINE_ void post() const {
		std::lock_guard lock(mutex);
		count++;
		condition.notify_one();
	}

	_ALWAYS_INLINE_ void wait() const {
		THREADING_NAMESPACE::unique_lock lock(mutex);
#ifdef DEBUG_ENABLED
		++awaiters;
#endif
		while (!count) { // Handle spurious wake-ups.
			condition.wait(lock);
		}
		--count;
#ifdef DEBUG_ENABLED
		--awaiters;
#endif
	}

	_ALWAYS_INLINE_ bool try_wait() const {
		std::lock_guard lock(mutex);
		if (count) {
			count--;
			return true;
		} else {
			return false;
		}
	}

#ifdef DEBUG_ENABLED
	~Semaphore() {
		// Destroying an std::condition_variable when not all threads waiting on it have been notified
		// invokes undefined behavior (e.g., it may be nicely destroyed or it may be awaited forever.)
		// That means other threads could still be running the body of std::condition_variable::wait()
		// but already past the safety checkpoint. That's the case for instance if that function is already
		// waiting to lock again.
		//
		// We will make the rule a bit more restrictive and simpler to understand at the same time: there
		// should not be any threads at any stage of the waiting by the time the semaphore is destroyed.
		//
		// We do so because of the following reasons:
		// - We have the guideline that threads must be awaited (i.e., completed), so the waiting thread
		//   must be completely done by the time the thread controlling it finally destroys the semaphore.
		//   Therefore, only a coding mistake could make the program run into such a attempt at premature
		//   destruction of the semaphore.
		// - In scripting, given that Semaphores are wrapped by RefCounted classes, in general it can't
		//   happen that a thread is trying to destroy a Semaphore while another is still doing whatever with
		//   it, so the simplification is mostly transparent to script writers.
		// - The redefined rule can be checked for failure to meet it, which is what this implementation does.
		//   This is useful to detect a few cases of potential misuse; namely:
		//   a) In scripting:
		//      * The coder is naughtily dealing with the reference count causing a semaphore to die prematurely.
		//      * The coder is letting the project reach its termination without having cleanly finished threads
		//        that await on semaphores (or at least, let the usual semaphore-controlled loop exit).
		//   b) In the native side, where Semaphore is not a ref-counted beast and certain coding mistakes can
		//      lead to its premature destruction as well.
		//
		// Let's let users know they are doing it wrong, but apply a, somewhat hacky, countermeasure against UB
		// in debug builds.
		std::lock_guard lock(mutex);
		if (awaiters) {
			WARN_PRINT(
					"A Semaphore object is being destroyed while one or more threads are still waiting on it.\n"
					"Please call post() on it as necessary to prevent such a situation and so ensure correct cleanup.");
			// And now, the hacky countermeasure (i.e., leak the condition variable).
			new (&condition) THREADING_NAMESPACE::condition_variable();
		}
	}
#endif
};

#else // No threads.

class Semaphore {
public:
	void post() const {}
	void wait() const {}
	bool try_wait() const {
		return true;
	};

#endif // THREADS_ENABLED

#endif // SEMAPHORE_H
```

RandomShaper reviewed (dismissed):
I'll do another pass later, once the existing comments have been addressed. As a general advice, I'd make all the no-threads versions of everything as slim as possible (commentless, without inline macros, etc.).

ChronoDK commented:
Did you test on iOS too? Doesn't work for me on iPhone 11. Loads and looks fine, but WebGL Context lost after a few seconds then "null is not an object (evaluating 'gl.getContextAttributes().antialias')

adamscott commented on core/os/mutex.h:
I had to edit your file in order for the project to compile.

adamscott commented on platform/web/detect.py:
```diff
@@ -74,6 +75,7 @@ def get_flags():
         # 100 KiB over -Os, which does not justify the negative impact on
         # run-time performance.
         ("optimize", "size"),
+        ("use_threads", False),
```
Considering that it's only for people self-compiling, I'll remove the new default value. We shouldn't be breaking compatibility/behavior until told to.

adamscott commented:
@jordo @akien-mga [I added CI](https://github.com/godotengine/godot/pull/85939/files#diff-1be8266b236042d967e952c14827382c7816d2dcffee44164616b57d53c9cabc) for the web platform non-threaded `target_release` build.

adamscott commented on core/io/ip.cpp:
```diff
@@ -129,6 +129,7 @@ IPAddress IP::resolve_hostname(const String &p_hostname, IP::Type p_type) {
 
 PackedStringArray IP::resolve_hostname_addresses(const String &p_hostname, Type p_type) {
 	List<IPAddress> res;
+
```
Will remove this superfluous line break.

RandomShaper commented on drivers/unix/thread_posix.cpp:
Maybe this one can even be excluded from the build at all. Alternatively, the `#ifdef` can enclose the whole file.

Regardless one or the other, my point is that OS-specific thread functions won't even be called at all and it may be more maintainable to just avoid calling the `init_thread_*()` functions.

RandomShaper commented on modules/raycast/raycast_occlusion_cull.cpp:
```diff
@@ -108,8 +108,12 @@ void RaycastOcclusionCull::RaycastHZBuffer::update_camera_rays(const Transform3D
 
 	debug_tex_range = td.z_far;
 
+#ifdef THREADS_ENABLED
```
It'd be more maintainable that the `WorkerThreadPool` acts as a dumb sequential runner of the tasks in cases like this, so there's no need to add these disjunctives every time.

adamscott replied:
I removed the `THREADS_ENABLED` here, as it was finally superfluous.

adamscott commented:
Rebased the branch PR in order to keep a single commit.

Krigu commented:
Thank you, this is awesome.

Not sure if it helps, but I tested the demo (https://adamscott.github.io/truck-town-demo-main-thread/) on multiple iOS Devices with Safari (iPhone, iPad and MacOS) and it seems to work fine on all of them. Only when starting the demo with Firefox 120.1 (on MacOS 14.0) the browser crashes.

ChronoDK commented:
I also tested a bit. On iPhone 11 Safari I had crashes at first, but the next day everything seemed to work fine. Maybe my browser was broken, or maybe you updated something. Either way, seems awesome now.

adamscott commented:
I'm having issues on my laptop, but I don't know if it's related to my build. I rather think that's related to rendering on my integrated graphics card.

![image](https://github.com/godotengine/godot/assets/270928/2e603616-2b34-401c-9286-586ddaa71041)

Here's the warning in text form:
> [!WARNING]
> WebGL warning: drawElementsInstanced: Depth texture comparison requests (e.g. `LINEAR`) Filtering, but behavior is implementation-defined, and so on some systems will sometimes behave as `NEAREST`. (warns once)

adamscott commented:
It seems that enabling sound is impacting performances. Here's a screenshot representative of the experience with sound, and another (a tad bit faster) without sound.

![Capture d’écran du 2023-12-14 20-13-49](https://github.com/godotengine/godot/assets/270928/4753a260-1cb7-4002-aa03-7459f17b3148)

![Capture d’écran du 2023-12-14 20-14-35](https://github.com/godotengine/godot/assets/270928/44f78c2b-cc98-4cf4-8fca-c9a28e85e747)

Currently, all the sound processing is done from within Godot. If we could decouple the processing of the sound from Godot and run it from Web Audio, maybe we could make audio run without (heavy) crackles (on low-end PCs and phones).

ChronoDK commented:
> It seems that enabling sound is impacting performances. Here's a screenshot representative of the experience with sound, 

Have you tried adjusting audio output latency? In Godot 3 that fixes most problems. I use output_latency.web=90
And also make sure sounds are preloaded (looks like they are here).

starry-abyss commented:
Regarding audio. Not sure maybe it's because my Mac Mini is low-end (it's from 2011), but in Godot 3 audio crackled for me there in the web version. At least the Ogg Vorbis music. Hopefully, this Web Audio thing helps!

popcar2 commented:
I've tried exporting using the no threads template on the windows editor and got this error: `Could not write file: index.audio.worklet.js`

Same error on 3 of my projects. Not sure if the current builds might have WIP changes that could be causing this.

![image](https://github.com/godotengine/godot/assets/16920817/d267a460-aad3-462b-94b8-d68e78f011c3)

AThousandShips commented:
Did you build the editor with this PR, or just the template?

popcar2 commented:
> Did you build the editor with this PR, or just the template?

I used the editor and template provided in this PR's artifacts.

AThousandShips commented:
You might not have write permissions for that folder, unsure though in this specific case

popcar2 commented:
It seems like I was using the template incorrectly, sorry! The zip artifact has other things bundled in it, I've extracted `.web_zip` from inside and used that as my template and it works just fine. I can confirm everything seems to be working without SharedArrayBuffers/CrossOriginIsolation. Performance seems good (although my projects are 2D and aren't very intensive).

I tried some Godot 4 projects on websites like Crazy Games, which notoriously never supported custom headers and all seems good. Thank you!

![826f65fb-3ede-4fc7-abcc-17d02b192daf](https://github.com/godotengine/godot/assets/16920817/d41cfa0e-1d58-4025-84e1-5e5e37e58daa)

AThousandShips commented:
Good a solution was found, hiding this as off topic as it wasn't an issue with this 🙂

daniel080400 commented:
Tested on iPad Air 4, iOS 15.7.1, it ran and works. But it's very laggy in the 3D scene (~6fps). 
Very great work to finally make Godot 4 HTML export work on iOS.

lostminds commented:
> Only when starting the demo with Firefox 120.1 (on MacOS 14.0) the browser crashes.

Testing (https://adamscott.github.io/truck-town-demo-main-thread/) on macOS 14.2.1 in Firefox 121.0 I'm not getting a crash and the game seems to be running as intended with good performance. It seems to get stuck loading (with the startup loading bar) for a while and then when loading the level after selecting the vehicle. And it's spitting out a bunch of these errors (1000x) during this time: 

```
Output buffer has not enough frames! Skipping output frame. 1000 index.js:15192:18
    onPrintError https://adamscott.github.io/truck-town-demo-main-thread/index.js:15192
    error https://adamscott.github.io/truck-town-demo-main-thread/index.js:9375
    onmessage https://adamscott.github.io/truck-town-demo-main-thread/index.js:9878
```
Maybe it's trying to flush the buffer, but since the game has locked up during these sequences, there's nothing to show to the screen?

akien-mga commented on platform/web/doc_classes/EditorExportPlatformWeb.xml:
```diff
@@ -47,6 +47,10 @@
 		<member name="variant/extensions_support" type="bool" setter="" getter="">
 		</member>
+		<member name="variant/thread_support" type="bool" setter="" getter="">
+			If enabled, the exported game will support threads. It requires [url=https://web.dev/articles/coop-coep]a "cross-origin isolated" website[/url], which can be difficult to setup and has brings some limitations (i.e. not being able to communicate with third-party websites).
```
```suggestion
			If enabled, the exported game will support threads. It requires [url=https://web.dev/articles/coop-coep]a "cross-origin isolated" website[/url], which can be difficult to setup and brings some limitations (e.g. not being able to communicate with third-party websites).
```

akien-mga commented on platform/web/doc_classes/EditorExportPlatformWeb.xml:
```diff
@@ -47,6 +47,10 @@
 		</member>
+		<member name="variant/thread_support" type="bool" setter="" getter="">
+			If enabled, the exported game will support threads. It requires [url=https://web.dev/articles/coop-coep]a "cross-origin isolated" website[/url], which can be difficult to setup and has brings some limitations (i.e. not being able to communicate with third-party websites).
+			If disabled, the exported game will not support threads. As a result, it is more prone to performance and audio issues, but will only require to be run on a HTTPS website.
```
I'm not even sure it needs to be on HTTPS if threads are disabled, AFAIK a secure context is also a requirement imposed by SharedArrayBuffer. Worth testing to confirm.

adamscott replied:
Yes, HTTPS is required for audio, at least.

adamscott replied:
See the requirements for AudioWorklet (needs HTTPS).
https://developer.mozilla.org/en-US/docs/Web/API/AudioWorklet

akien-mga commented on platform/web/js/engine/features.js:
```diff
@@ -84,11 +84,16 @@ const Features = { // eslint-disable-line no-unused-vars
+		const compiledWithThreads = ___GODOT_THREADS_ENABLED;
+		if (compiledWithThreads) {
+			if (!Features.isCrossOriginIsolated()) {
+				missing.push('Cross Origin Isolation - Check web server configuration (send correct headers)');
```
```suggestion
				missing.push('Cross-Origin Isolation - Check that the web server configuration sends the correct headers.');
```

akien-mga commented on platform/web/js/engine/features.js:
```diff
@@ -84,11 +84,16 @@ const Features = { // eslint-disable-line no-unused-vars
+				missing.push('Cross Origin Isolation - Check web server configuration (send correct headers)');
+			}
+			if (!Features.isSharedArrayBufferAvailable()) {
+				missing.push('SharedArrayBuffer - Check web server configuration (send correct headers)');
```
```suggestion
				missing.push('SharedArrayBuffer - Check that the web server configuration sends the correct headers.');
```

akien-mga reviewed (approved):
Looks pretty good to me!

I think further work on audio performance, etc. should be left to follow-up PRs, this is a good first step to add back support for compiling the engine without thread support.

Faless commented on platform/web/js/engine/features.js:
```diff
@@ -84,11 +84,16 @@ const Features = { // eslint-disable-line no-unused-vars
-			missing.push('SharedArrayBuffer - Check web server configuration (send correct headers)');
+
+		// eslint-disable-next-line no-undef
+		const compiledWithThreads = ___GODOT_THREADS_ENABLED;
```
We should have `getMissingFeatures` take a parameter, and do the substitution in the html file.
```
	getMissingFeatures: function (wantsThreads = true) {
		// ...
		if (wantsThreads) {
			// ...
		}
		// ...
	},
```

Faless reviewed (commented):
Looking great :rocket:  ,
I realize I had a few pending comments which were never published :(.

adamscott commented on SConstruct:
```diff
@@ -964,6 +965,9 @@ if selected_platform in platform_list:
         env.Tool("compilation_db")
         env.Alias("compiledb", env.CompilationDatabase())
 
+    if env["use_threads"]:
+        env.Append(CPPDEFINES=["THREADS_ENABLED"])
```
Did it! [Line 948-949](https://github.com/godotengine/godot/pull/85939/files#diff-d78c5c377c03dd251edc72cdb0ae7d9986fa08361eb392187218f8c670c0bc4eR948-R949) now.

Smooth-E commented:
In contrary to [popcar2's results](https://github.com/godotengine/godot/pull/85939#issuecomment-1866467469) I am unable to play the game on Crazy Games after exporting it with both the editor (for Windows) and the template from [this](https://github.com/adamscott/godot/actions/runs/7465563648) GitHub Action. I tried using both of the following files as templates, but none of them work giving me the following message.

Files I tried as a template:
```txt
web-nothreads-template.zip/godot.web.template_release.wasm32.nothreads.zip
```
```txt
web-nothreads-template.zip/.web_zip (packed back as a zip)
```

The message I am getting on Crazy Games:
![image](https://github.com/godotengine/godot/assets/71522503/596c1ad8-979a-4330-ad11-2ce7de47b637)

I tried setting the template files as *Custom template/Release* in project export settings and I also tried moving them to Godot's export templates folder in `AppData`.
```txt
C:\Users\Username\AppData\Roaming\Godot\export_templates\4.3.dev\web_nothreads_release.zip
```


Here are some of my export settings:
- *Compile with debug* option is unchecked, as this seems to be a release template.
- *Variant/Thread support* is unchecked in export settings
- *Features/Feature list* is `etc2, s3tc, wasm32, web`, no custom features
- *Resources/Filters to export non-project files* is `*txt`
- *PWA* is disabled

AThousandShips commented on core/os/semaphore.h:
```diff
@@ -130,4 +132,17 @@ class Semaphore {
+	void wait() const {}
+	bool try_wait() const {
+		return true;
+	};
```
```suggestion
	}
```

AThousandShips commented on core/os/mutex.h:
```diff
@@ -125,7 +127,7 @@ class MutexLock {
 public:
-	_ALWAYS_INLINE_ explicit MutexLock(const MutexT &p_mutex) :
+	explicit MutexLock(const MutexT &p_mutex) :
 			lock(p_mutex.mutex){};
```
```suggestion
			lock(p_mutex.mutex) {}
```

AThousandShips commented on main/main.cpp:
```diff
@@ -1625,6 +1626,9 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 		}
+#else
+		WorkerThreadPool::get_singleton()->init(0, false, 0);
+#endif
```
```suggestion
#endif // THREADS_ENABLED
```

adamscott replied:
It would be the only macro to do so in the whole file. :confused:

AThousandShips commented on core/os/thread.cpp:
```diff
@@ -106,5 +109,6 @@ Thread::~Thread() {
 		thread.detach();
 	}
 }
+#endif
```
```suggestion
#endif // THREADS_ENABLED
```

adamscott replied:
Thanks! I completely forgot to add this one.

AThousandShips commented on core/object/worker_thread_pool.cpp:
```diff
@@ -184,6 +190,7 @@ void WorkerThreadPool::_process_task(Task *p_task) {
 			task_available_semaphore.post();
 		}
 	}
+#endif
```
```suggestion
#endif // THREADS_ENABLED
```

adamscott replied:
Not needed here.

AThousandShips commented on core/object/worker_thread_pool.cpp:
```diff
@@ -80,6 +84,7 @@ void WorkerThreadPool::_process_task(Task *p_task) {
 		}
 		task_mutex.unlock();
 	}
+#endif
```
```suggestion
#endif // THREADS_ENABLED
```

adamscott replied:
Not needed here.

AThousandShips reviewed (approved):
LGTM, just some style nits

popcar2 commented:
> In contrary to [popcar2's results](https://github.com/godotengine/godot/pull/85939#issuecomment-1866467469) I am unable to play my game on Crazy Games after exporting it with both the editor (for Windows) and the template from [this](https://github.com/adamscott/godot/actions/runs/7465563648) GitHub Action. I tried using both of the following files as templates, but none of them work

Good catch. I can confirm that using the newest windows build and newest no_threads release template asks for SharedArrayBuffers even when thread support is **off**. 

New no_threads template export:

![firefox_N4CEzdiRYp](https://github.com/godotengine/godot/assets/16920817/bdb18620-9bd2-4fb0-baf2-2860f6ab44eb)

Old no_threads template export (exactly the same settings):

![firefox_ryHMBdyWJU](https://github.com/godotengine/godot/assets/16920817/4f7774ff-0b7e-4f9c-887c-b94b04fc0d1c)

This might just be an issue with GitHub CI compiling the export wrong, but I wanted to confirm anyways. Here's the old template I used that worked just fine for me:

[Godot Threadless Template Test.zip](https://github.com/godotengine/godot/files/13902461/Godot.Threadless.Template.Test.zip)

Smooth-E commented:
Thanks, @popcar2! Indeed, using an older template you provided with the editor from the GitHub Action mentioned by me above allows to play the game on Crazy Games.

Smooth-E commented:
Comparing the template provided by popcar2 and the template I got from CI, here is the part which is pushing the text of an error (probably). `supportsThreads` is set to `true` here for some reason. I do not really know how templates in Godot works but changing this to `false` in an already exported project does not solve the issue.

![image](https://github.com/godotengine/godot/assets/71522503/0869d2f8-ef87-4f5f-a842-7119a7c38f82)

Don't mind VSCode's GitLens comments, I've just thrown one version of template on top of another and compared them with GitLens.

adamscott commented:
> ```
> Output buffer has not enough frames! Skipping output frame. 1000 index.js:15192:18
>     onPrintError https://adamscott.github.io/truck-town-demo-main-thread/index.js:15192
>     error https://adamscott.github.io/truck-town-demo-main-thread/index.js:9375
>     onmessage https://adamscott.github.io/truck-town-demo-main-thread/index.js:9878
> ```
> 
> 
>     
>   
> 
> Maybe it's trying to flush the buffer, but since the game has locked up during these sequences, there's nothing to show to the screen?

Makes me think that we should remove these alerts. I'll add this to my todo list for this PR.

akien-mga commented on core/os/os.cpp:
```diff
@@ -504,6 +504,12 @@ bool OS::has_feature(const String &p_feature) {
 	}
 #endif
 
+#ifdef THREADS_DEFINED
```
Wrong define.

adamscott replied:
😨

akien-mga commented on main/main.cpp:
```diff
@@ -1840,13 +1846,13 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 	}
 
-	// Start with RenderingDevice-based backends.
+// Start with RenderingDevice-based backends.
```
Which clang-format version do you have? It shouldn't do these changes.

akien-mga replied:
I tested with clang-format 15, it seems to be fine with both indentations. I think it makes more sense to keep at the code scope level.

adamscott replied:
Done. I don't know what happened.

imvenx commented:
I'm trying to run godot with this version and export to web, but I'm missing the template, and 4.2.1.stable is not working as a substitute
![image](https://github.com/godotengine/godot/assets/115882542/f86476db-34b9-46bf-88f4-d77dca2880a8)
![image](https://github.com/godotengine/godot/assets/115882542/45fe6f0d-cf7a-4b6f-94de-78b462f605a9)

imvenx commented:
trying to run it but getting this

![image](https://github.com/godotengine/godot/assets/115882542/dcf9f80e-9672-42ac-a1f5-4bed550a35ae)

Smooth-E commented:
> trying to run web export on godot editor but getting this
> 
> ![image](https://private-user-images.githubusercontent.com/115882542/296083861-dcf9f80e-9672-42ac-a1f5-4bed550a35ae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUwNTIyOTIsIm5iZiI6MTcwNTA1MTk5MiwicGF0aCI6Ii8xMTU4ODI1NDIvMjk2MDgzODYxLWRjZjlmODBlLTk2NzItNDJhYy1hMWY1LTRiZWQ1NTBhMzVhZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExMlQwOTMzMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NWUwYjAwNTMxMDAwMmQyYTE1NGNlOWY1NTYxNDM4NmE3OTQ0MzVkMTUwNmZiOGVlMjg2NmIwNzVhYWQzN2MzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.6zUWy8GUEk_sQKN4lvhq2vfEDw34p61AlvM9U3UD5yQ)

@imvenx, here is what you should do to fix this. In case you are doing it later, when the template is fixed, then you can get the template from [Adam's repository GitHub Actions](https://github.com/adamscott/godot/actions) (on an appropriate branch, `single-threaded-godot-4`). See [this](https://github.com/godotengine/godot/pull/85939#issuecomment-1886943816) comment for issue description with the current template (from the latest Action as of mentioned comment). For an older, but working template, download the zip file [mentioned by popcar2 (link at the bottom)](https://github.com/godotengine/godot/pull/85939#issuecomment-1886943816).

If you are getting the template from GitHub Actions, download `web-nothreads-template` under artifacts and extract `godot.web.template_release.wasm32.nothreads.zip` from inside of it.

Now, under your project's export settings, select this zip file as a custom release template. When choosing the output directory (after you click *Export project...*), uncheck *Build with debug* at the bottom. Alternatively you can move mentioned zip file to your AppData directory (full path on your screenshot) as a release template.

I recommend first trying the template from Actions and then the older one by popcar2. Especially if you will do this after some new commits are pushed.

akien-mga commented on SConstruct:
```diff
@@ -170,23 +170,68 @@ opts = Variables(customs, ARGUMENTS)
 opts.Add("p", "Platform (alias for 'platform')", "")
-opts.Add(EnumVariable("target", "Compilation target", "editor", ("editor", "template_release", "template_debug")))
-opts.Add(EnumVariable("arch", "CPU architecture", "auto", ["auto"] + architectures, architecture_aliases))
+opts.Add(
+    EnumVariable(
+        "target",
+        "Compilation target",
+        "editor",
+        ("editor", "template_release", "template_debug"),
+    )
+)
```
Sounds like you ran `black` without `-l 120`, so it used a lower limit for the line length. Those changes should be reverted.

adamscott replied:
Done! Thanks!

akien-mga commented on SConstruct:
```diff
@@ -170,23 +170,68 @@ opts = Variables(customs, ARGUMENTS)
+        False,
+    )
+)
+opts.Add(BoolVariable("use_threads", "Enable threading support", True))
```
I wanted to test this PR and intuitively I used `scons threads=no`. We still have a couple `use_<thing>` options but I think for most we're now using just `<thing>=yes|no`.

WDYT? I don't feel too strongly about this either way. Just noticed that `threads=no` came more naturally to me.

adamscott replied:
Renamed to `threads=yes|no`!

adamscott commented:
Renamed `use_threads=yes|no` `threads=yes|no`, following @akien-mga's interrogation.

adamscott commented:
@Smooth-E I had errors while trying to export [Truck Town](https://github.com/godotengine/godot-demo-projects/tree/master/3d/truck_town) using the GitHub web no-threads artifiact, but I had no issues if I build it myself. So my PR seems fine, at least the Godot part.

adamscott commented:
@Smooth-E I'm a goofball, I used the artifact wrong. I used it directly as the template instead of unzipping it. 

I tested the artifact template locally and it works! Could you test again?

adamscott commented:
Added the known issue that sound is broken on non-web platforms.

Smooth-E commented:
> @Smooth-E I'm a goofball, I used the artifact wrong. I used it directly as the template instead of unzipping it.
> 
> I tested the artifact template locally and it works! Could you test again?

Strange. I got both the editor and the template from [this GitHub Action](https://github.com/adamscott/godot/actions/runs/7504605083) and on Crazy Games, no matter the file I use as a template, the error is still present.

Here is what I got in browser's console (game-specific tokens are omitted):
```
Error
The following features required to run Godot projects on the Web are missing:
Cross-Origin Isolation - Check that the web server configuration sends the correct headers.
SharedArrayBuffer - Check that the web server configuration sends the correct headers. index.html:213:11
    displayFailureNotice https://prod-dpgames.crazygames.com/<game-specific identifier>/index.html?<arguments>:213
    <anonymous> https://prod-dpgames.crazygames.com/<game-specific identifier>/index.html?<arguments>:224
    <anonymous> https://prod-dpgames.crazygames.com/<game-specific identifier>/index.html?<arguments>:247
```

Files I tried as templates are those (same paths as in my [previous attempt](https://github.com/godotengine/godot/pull/85939#issuecomment-1886300794):
```
web-nothreads-template.zip/godot.web.template_release.wasm32.nothreads.zip
```
```
web-nothreads-template.zip/.web_zip (packed back as a zip)
```

Got to say that using [popcar2's template](https://github.com/godotengine/godot/pull/85939#issuecomment-1886943816) still works fine.

~~Maybe try publishing on Crazy Games specifically? Registration there does not require any personal information and verification other than email. You can launch and demo the game in a test sandbox right away.~~

I tried launching the exported game locally with `python -m http.server` and got the same error as on Crazy Games:
```
Error
The following features required to run Godot projects on the Web are missing:
Cross-Origin Isolation - Check that the web server configuration sends the correct headers.
SharedArrayBuffer - Check that the web server configuration sends the correct headers. localhost:8000:213:11
    displayFailureNotice http://localhost:8000/:213
    <anonymous> http://localhost:8000/:224
    <anonymous> http://localhost:8000/:247
```

adamscott commented:
> I tried launching the exported game locally with `python -m http.server` and got the same error as on Crazy Games:
> 
> ```
> Error
> The following features required to run Godot projects on the Web are missing:
> Cross-Origin Isolation - Check that the web server configuration sends the correct headers.
> SharedArrayBuffer - Check that the web server configuration sends the correct headers. localhost:8000:213:11
>     displayFailureNotice http://localhost:8000/:213
>     <anonymous> http://localhost:8000/:224
>     <anonymous> http://localhost:8000/:247
> ```

Weird. With my build, you're supposed to have an `HTTPS` error while running with `python -m http.server`, as HTTPS is required.

![image](https://github.com/godotengine/godot/assets/270928/60a4d8de-7b74-411e-a327-cbf6619b8a98)

Are you sure you're not testing an old exported game instead?

Smooth-E commented:
> Are you sure you're not testing an old exported game instead?

So, again. I've got both the Windows editor and `web-nothreads-template` from [this GitHub Action](https://github.com/adamscott/godot/actions/runs/7504605083). I created a new project with this editor. An `AppData\Roaming\Godot\export_templates\4.3.dev` folder is empty for me. I used both a zipped file and `.web_zip` as custom release templates and in both cases I get CORS + SharedArrayBuffer errors when serving with `python -m http.server`.

I am not getting any https-related errors and I have no crossed lock icon in Firefox' search bar (seen on your screenshot) when I visit `localhost:8000`.

Are you using builds from CI or ones compiled locally by yourself? May it be Windows-related? I can try running it on a Linux VM. Perhaps, someone else should test it (on Windows), since it may be related to my configuration somehow 🤷

akien-mga commented:
I tested the PR with the `Web / Template w/o threads` artifact, and it works for me.

@Smooth-E For it to work, you need to make sure you're using not only the template built from this PR, **but also the editor built from this PR.** It won't work with a custom build from `master`, a dev snapshot, or 4.2.1.stable, as this PR includes editor code needed to export the template with the proper configuration. The `Windows / Editor` artifact should do the trick if you're on Windows. (For me on Linux I had to compile from source, as the only "normal" Linux editor built is with .NET support, so it won't let me export to Web.)

I configured things like this, using the same template for both Debug and Release templates:

![image](https://github.com/godotengine/godot/assets/4701338/57f0602e-8651-4829-89bb-145c0c907794)

It should work by putting it only for Release and doing a release export, but I put it in Debug too to be able to test the one-click deploy from the editor. The one-click deploy failed at spawning the server for me, but I don't think it's linked to this PR (or it might be because I built a `nothreads` Linux editor build for the lulz, which isn't really supported).

With a normal release export, I hosted it with `python -m http.server`.
On `http://0.0.0.0:8000`, I get the error about lack of HTTPS, as expected. On `localhost:8000`, it works.

![image](https://github.com/godotengine/godot/assets/4701338/cf39613f-f53a-4fa0-a57e-b0b7d300b1c4)

![image](https://github.com/godotengine/godot/assets/4701338/3cfe045b-c584-4878-a3c3-bd7ccb9e5fec)

akien-mga commented:
@adamscott There's an issue with One-Click Deploy:

```
Uncaught ReferenceError: $GODOT_THREADS_ENABLED is not defined
    <anonymous> http://localhost:8060/tmp_js_export.html:141
```

To test, add a "runnable" export preset and configure the debug template to point to one made from this PR.

Calinou commented:
> Invalid mix rate of 0, consider reassigning setting 'audio/driver/mix_rate'.

You can ignore this message as a mix rate of `0` is valid on the web platform (it lets the browser choose).

adamscott commented:
> > Invalid mix rate of 0, consider reassigning setting 'audio/driver/mix_rate'.
> 
> You can ignore this message as a mix rate of `0` is valid on the web platform (it lets the browser choose the most suitable mix rate automatically).

I could mask this error from WEB_ENABLED, though.

adamscott commented:
> @adamscott There's an issue with One-Click Deploy:
> 
> ```
> Uncaught ReferenceError: $GODOT_THREADS_ENABLED is not defined
>     <anonymous> http://localhost:8060/tmp_js_export.html:141
> ```
> 
> To test, add a "runnable" export preset and configure the debug template to point to one made from this PR.

AH. I understand now why it was working for me. I was only testing with release templates, and the "run in browser" build was using an old (working) debug template.

adamscott commented:
@akien-mga I'm not able to replicate your issue, even with the debug template build up-to-date.

I tried erasing non-threads templates, setting the templates in the export settings. But clicking One-click deploy for me works without hiccups for me, even on Firefox or Chrome Incognito mode.

akien-mga commented:
> @akien-mga I'm not able to replicate your issue, even with the debug template build up-to-date.
> 
> I tried erasing non-threads templates, setting the templates in the export settings. But clicking One-click deploy for me works without hiccups for me, even on Firefox or Chrome Incognito mode.

I tested the current commit (7c4cb14a5ed02a3b1032728ca726dae9fd58faac) and it seems to be working fine for me.

I tested:
- Custom built Godot editor on Linux
- Custom built debug template (`scons p=web target=template_debug`)
- Release template from the GitHub Actions used as Custom debug template

Both my custom built and the GH Actions artifact seem to be working fine.

akien-mga reviewed (approved):
I think this is good to go!

akien-mga commented:
Wait I spoke too fast again...

The template from GHA still doesn't work for me. It still tries to enable SharedArrayBuffer.

And my locally built templates lacked `threads=no` so they're not a good test :P I'll rebuild.

akien-mga commented:
Yep, local build with `scons p=web target=template_debug verbose=yes scu_build=yes threads=no`, I still have this error:

![image](https://github.com/godotengine/godot/assets/4701338/37a73805-f479-4025-8d92-77f6687b053d)

So this PR is not working yet for me.

For the record, the One-click deploy mode isn't a good way to test the no-threads builds, as the server we spawn has the required headers for SharedArrayBuffer.

So to be extremely clear, here's what I'm testing:

- Compile Linux editor from 7c4cb14a5ed02a3b1032728ca726dae9fd58faac with `scons p=linuxbsd dev_build=yes dev_mode=yes linker=mold scu_build=all` (`dev_*`, `linker`, `scu_build` shouldn't matter here, but that's just what I use for faster builds)
- Compile Web debug template **without thread support** with `scons p=web target=template_debug verbose=yes scu_build=yes threads=no` 
- Edit https://github.com/HarmonyHoney/CandyWrapper with the editor build I made
- Add Web export template, configure the Custom debug template path to `~/godot/folder/bin/godot.web.template_debug.wasm32.nothreads.zip`
- Disable the "Thread Support" checkbox

![image](https://github.com/godotengine/godot/assets/4701338/fb8d61fb-798e-47fb-9a24-d00bcbe7cc05)

- "Export Project...", "Export With Debug"
- In the export folder, `python -m http.server`
- Browse `localhost:8000`

Same issue if I use the export template artifact from GitHub Actions instead of my custom build.

adamscott commented:
@akien-mga @Smooth-E @popcar2 @wojtekmal I found and fixed the bug.

```cpp
#ifdef THREADS_ENABLED
	replaces["$GODOT_THREADS_ENABLED"] = "true";
#else
	replaces["$GODOT_THREADS_ENABLED"] = "false";
#endif
```

That code was setting the requirements for threads as `true` if the editor itself had threads support. `false` otherwise. That bug explains why the template from my Github actions weren't working.

I replaced that code with the actual value checked in the editor:
```cpp
	if (p_preset->get("variant/thread_support")) {
		replaces["$GODOT_THREADS_ENABLED"] = "true";
	} else {
		replaces["$GODOT_THREADS_ENABLED"] = "false";
	}
```

akien-mga commented:
Great, that would explain why it worked for me last week when I tested with a Linux editor build with threads disabled, just for fun.

I'll retest today and merge.

akien-mga reviewed (approved):
Tested successfully! And I don't spot any other misplaced `TOOLS_ENABLED` check where it's meant to check the target config, so we should be good to go finally.

akien-mga commented:
Thanks! Amazing work :tada: :1st_place_medal:

jordo commented:
nice work!   paving the way for our team to eventually move to 4.X!  Thanks @adamscott

geekley commented:
Hi, I came from [this post](https://godotengine.org/article/progress-report-web-export-in-4-3/), which said to report errors here.
> If you encounter any issue while testing my PR or while testing out the single-threaded samples builds, please report them here.

I tried the [catburglar-main-thread-samples](https://adamscott.github.io/catburglar-main-thread-samples/) on Firefox. It worked fine until completing the (1st?) level, when it stopped responding. That's after getting at least $300 and returning to the entrance.
The browser console logged this error when the game tried to exit that scene/stage after completing the goal:
```
Uncaught TypeError: AudioNode.connect: Argument 1 is not valid for any of the 2-argument overloads.
    pause https://adamscott.github.io/catburglar-main-thread-samples/index.js:9944
    pauseSampleNode https://adamscott.github.io/catburglar-main-thread-samples/index.js:9814
    sample_set_pause https://adamscott.github.io/catburglar-main-thread-samples/index.js:10380
    _godot_audio_sample_set_pause https://adamscott.github.io/catburglar-main-thread-samples/index.js:10559
    callUserCallback https://adamscott.github.io/catburglar-main-thread-samples/index.js:6639
    runIter https://adamscott.github.io/catburglar-main-thread-samples/index.js:6714
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6600
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
    requestAnimationFrame https://adamscott.github.io/catburglar-main-thread-samples/index.js:7009
    Browser_mainLoop_scheduler_rAF https://adamscott.github.io/catburglar-main-thread-samples/index.js:5893
    Browser_mainLoop_runner https://adamscott.github.io/catburglar-main-thread-samples/index.js:6613
index.js:9944:19
```

JunShiozawa commented:
@adamscott 
Could you please backport this PR(samples) to Godot 3.x?
I need WebGL 1.0

adamscott commented:
@JunShiozawa This isn't the samples PR. This one is: #91382
But it will not be ported by me, at least. But if somebody wants to port the feature, they're welcome to do so!

JunShiozawa commented:
It doesn't work.

Godot 4.3 stable(Linux)

Web export

Thread Support Off

BGM and SE Platback Type are "Sample"

Safari

iPad OS

Safari error message

null is not an object (evaluating 'gl.getContextAttributes().antialias')


It works on Chrome for Android.

It works on Firefox for Android.

JunShiozawa commented:
It doesn't work.

Godot 4.3 stable(Linux)

Web export

Thread Support Off

BGM and SE Playback Type are "Sample"

Safari

iPad OS

Safari error message

null is not an object (evaluating 'gl.getContextAttributes().antialias')

It works on Chrome for Android.

It works on Firefox for Android.

akien-mga commented:
@JunShiozawa Please open an issue with steps to reproduce and a reproduction project.

JunShiozawa commented:
@akien-mga 

https://docs.godotengine.org/ja/4.x/tutorials/export/exporting_for_web.html

> See Can I use WebGL 2.0 for a list of browser versions supporting WebGL 2.0. Note that Safari has several issues with WebGL 2.0 support that other browsers don't have, so we recommend using a Chromium-based browser or Firefox if possible.

In the end, due to the bug in Safari's WebGL 2.0, I will have to continue using Godot 3.x (WebGL 1.0 GLES2).

JunShiozawa commented:
@akien-mga 

https://docs.godotengine.org/ja/4.x/tutorials/export/exporting_for_web.html

>See Can I use WebGL 2.0 for a list of browser versions supporting WebGL 2.0. Note that Safari has several issues with WebGL 2.0 support that other browsers don't have, so we recommend using a Chromium-based browser or Firefox if possible.

In the end, due to the bug in Safari's WebGL 2.0, I will have to continue using Godot 3.x(WebGL 1.0 GLES2).

JunShiozawa commented:
@akien-mga 

https://docs.godotengine.org/ja/4.x/tutorials/export/exporting_for_web.html

> See Can I use WebGL 2.0 for a list of browser versions supporting WebGL 2.0. Note that Safari has several issues with WebGL 2.0 support that other browsers don't have, so we recommend using a Chromium-based browser or Firefox if possible.

Not only are there problems with SharedArrayBuffer, Thread Support, and Playback Type "Sample", but Safari's WebGL 2.0 is buggy to begin with, and in the end, because of the bugs in Safari's WebGL 2.0, we will have to continue using Godot 3.x (WebGL 1.0 GLES2) in the future.

JunShiozawa commented:
@akien-mga 

https://docs.godotengine.org/ja/4.x/tutorials/export/exporting_for_web.html

> See Can I use WebGL 2.0 for a list of browser versions supporting WebGL 2.0. Note that Safari has several issues with WebGL 2.0 support that other browsers don't have, so we recommend using a Chromium-based browser or Firefox if possible.

Not only are there problems with SharedArrayBuffer, Thread Support, and Playback Type "Sample", but Safari's WebGL 2.0 is buggy to begin with, and in the end, because of the bugs in Safari's WebGL 2.0, we will have to continue using Godot 3.x(WebGL 1.0 GLES2) in the future.

ChronoDK commented:
4.x games do work on iOS Safari, but only for a while. Simple games will work for a long time, but even they will eventually crash/reload the browser. No idea why. It happens with any project, and more often with a complex project.

akien-mga commented:
@JunShiozawa Ok, but there's nothing actionable then, we can't fix Safari. If you have concrete bugs to report with it, please open bug reports with reproduction projects and we can see if we can workaround Safari shortcomings. If not, yes you can keep using 3.x for better compatibility.
